File: unpack_models.js
'use strict';

const AWS = require('aws-sdk');
const s3 = new AWS.S3();
const tar = require('tar-stream');
const zlib = require('zlib');
const stream = require('stream');
const uuidv4 = require('uuid/v4');

/*
 Expects s3:ObjectCreated event which is fired when SageMaker creates model artifcats.
 Expected S3 key should be like: 'transformed_models/<project_name>/<model_name>/test-mindful-messages10-202005112319-83072c36cecb-f/output/model.tar.gz.out'
 */
module.exports.unpack = function(event, context) {
  
  console.log(`processing event records from firehose bucket SNS event ${JSON.stringify(event)}`);

  /*
   We expect only 1 s3 event record, see forums:
   https://forums.aws.amazon.com/thread.jspa?messageID=592264#592264
   */
  if (!event.Records[0].s3) {
    console.log(`WARN: Invalid S3 event ${JSON.stringify(event)}`);
    context.fail();
    return;
  }
  
  let s3Record = event.Records[0].s3;
  const inputKey = s3Record.object.key;
  const pathParts = inputKey.split('/');
  if (pathParts[pathParts.length - 1] != 'model.tar.gz.out') {
    console.log(`WARN: Invalid S3 event ${JSON.stringify(event)}`);
    context.fail();
    return;
  }

  const [baseDir, projectName, modelName] = pathParts;

  var params = {
      Bucket: s3Record.bucket.name,
      Key: inputKey
  };

  var dataStream = s3.getObject(params).createReadStream();

  var extract = tar.extract();

  extract.on('entry', function(header, inputStream, next) {
      const extension = getExtension(header.name)
      
      if (extension) {
        const key = getTimestampedS3Key(projectName, modelName, extension);
        const latestKey = getLatestS3Key(projectName, modelName, extension);
        inputStream.pipe(uploadFromStream(key, latestKey));
  
        inputStream.on('end', function() {
            next(); // ready for next entry
        });
      } else {
        next();
      }

      inputStream.resume(); // just auto drain the stream
  });

  extract.on('finish', function() {
      // all entries read
  });

  dataStream.pipe(zlib.createGunzip()).pipe(extract);
}

function uploadFromStream(key, latestKey) {
    var pass = new stream.PassThrough();

    var writeParams = {
        Bucket: process.env.MODELS_BUCKET,
        Key: key,
        Body: pass
    };
    
    if (key.endsWith(".mlmodel")) {
      writeParams.ContentType = "application/protobuf" // allows cloudfront to automatically compress .mlmodel files
    } else if (key.endsWith(".gz")) {
      writeParams.ContentType = "application/gzip" // allows clients to automatically decompress
    }

    const copyParams = {
      Bucket: process.env.MODELS_BUCKET,
      CopySource: process.env.MODELS_BUCKET + '/' + key,
      Key: latestKey
    };

    console.log(`uplading with params ${JSON.stringify(writeParams)}`)
    s3.upload(writeParams).promise().then((data) => {
      console.log(`copying with params ${JSON.stringify(copyParams)}`)
      return s3.copyObject(copyParams).promise();
    }).then((data) => {
      console.log('copy completed');
    }, (err) => {
      console.log('copy failed', err);
    });

    return pass;
}

function getExtension(filePath) {
  if (filePath.endsWith('.mlmodel')) {
    return '.mlmodel'
  } else if (filePath.endsWith('.mlmodel.gz')) {
    return '.mlmodel.gz'
  } else if (filePath.endsWith('.xgb.gz')) {
    return '.xgb.gz'
  } else {
    return null;
  }
}

function getLatestS3Key(projectName, modelName, extension) {
  return `models/${projectName}/latest/improve-${modelName}${extension}`
}

function getTimestampedS3Key(projectName, modelName, extension) {
  const now = new Date();
  let dateStr = `${now.getUTCFullYear()}-${now.getUTCMonth()}-${now.getUTCDate()}-`;
  dateStr += `${now.getUTCHours()}-${now.getUTCMinutes()}-${now.getUTCSeconds()}`;

  const uuidStr = uuidv4();

  return `models/${projectName}/archive/improve-${modelName}-${dateStr}-${uuidStr}.${extension}`
}

File: src/train/unpack_models.py
import boto3
from io import BytesIO
from datetime import datetime
import os
import tarfile
import gzip
from uuid import uuid4

import src.train.constants as tc


def unpack(event, context):
    s3_client = boto3.client('s3')

    if not event['Records'][0]['s3']:
        raise TypeError('Invalid S3 event: {}'.format(event))

    s3_record = event['Records'][0]['s3']

    s3_bucket = s3_record['bucket']['name']
    s3_key = s3_record['object']['key']
    
    path_parts = s3_key.split('/')

    if path_parts[-1] != tc.EXPECTED_TRAINER_OUTPUT_FILENAME:
        raise ValueError(
            'Invalid S3 event - model filename `{}` differs from expected `{}`'
            .format(s3_key, tc.EXPECTED_TRAINER_OUTPUT_FILENAME))

    s3_model_name = path_parts[2]

    print(f'loading s3://{s3_bucket}/{s3_key}')
    models_object = s3_client.get_object(Bucket=s3_bucket, Key=s3_key)

    if 'Body' not in models_object:
        raise ValueError('Missing `Body` in S3 `get_object()` response')

    # models_object['Body'] is a 'botocore.response.StreamingBody'
    # need to call read() on it to access the *.tar.gz file
    tar_gz_models_buffer = BytesIO(models_object['Body'].read())

    with tarfile.open(fileobj=tar_gz_models_buffer) as tar_gz_models_file:
        filenames = tar_gz_models_file.getnames()

        for filename in filenames:

            if filename == 'model.xgb':
                upload_extension = '.xgb.gz'
            elif filename == 'model.mlmodel':
                upload_extension = '.mlmodel.gz'
            else:
                print(f'skipping {filename}')
                continue

            key = get_timestamped_s3_key(model_name=s3_model_name, extension=upload_extension)
            latest_key = get_latest_s3_key(model_name=s3_model_name, extension=upload_extension)

            model_member = tar_gz_models_file.getmember(filename)
            model = tar_gz_models_file.extractfile(model_member).read()

            upload_model(
                key=key, latest_key=latest_key, s3_client=s3_client,
                model_fileobject=model)


def upload_model(key: str, latest_key: str, s3_client, model_fileobject: BytesIO):

    models_bucket = os.getenv(tc.MODELS_BUCKET_ENVVAR)
    
    # upload gzipped compressed model
    write_params = {
        'Fileobj': BytesIO(gzip.compress(model_fileobject)),
        'Bucket': models_bucket,
        'Key': key,
        'ExtraArgs': {
            'ContentType': 'application/gzip'
        }
    }

    copy_params = {
        'Bucket': models_bucket,
        'CopySource': models_bucket + '/' + key,
        'Key': latest_key,
    }

    print(f'writing s3://{models_bucket}/{key}')
    s3_client.upload_fileobj(**write_params)

    print(f'copying to s3://{models_bucket}/{latest_key}')
    s3_client.copy_object(**copy_params)

def get_latest_s3_key(model_name: str, extension: str) -> str:
    return f'models/latest/{model_name}{extension}'

def get_timestamped_s3_key(
        model_name: str, extension: str, model_uuid: str = None) -> str:
    date_str = datetime.now().strftime('%Y-%m-%d-%H-%M-%S')

    return f'models/archive/{model_name}/{model_name}-{date_str}-{uuid4()}{extension}'
